include/rsl/parameter_validators.hpp

```diff
@@ -0,0 +1,301 @@
+    return detail::size_cmp<T>(parameter, size, "greater than", std::greater<size_t>());
+}
+
+/**
+ * @brief Is the array size of parameter less than passed in size?
+ * @pre rclcpp::Parameter must be an array type
+ * @tparam T the interior type of array; Example: for parameter type double_array, T = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto size_lt(rclcpp::Parameter const& parameter, size_t const size) {
+    return detail::size_cmp<T>(parameter, size, "less than", std::less<size_t>());
+}
+
+/**
+ * @brief Is the size of the value passed in not zero?
+ * @pre rclcpp::Parameter must be an array type or a string
+ * @tparam T the interior type of array or std::string; Example: for parameter type double_array, T
+ * = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto not_empty(rclcpp::Parameter const& parameter)
+    -> tl::expected<void, std::string> {
+    switch (parameter.get_type()) {
+        case rclcpp::ParameterType::PARAMETER_STRING:
+            if (auto param_value = parameter.get_value<std::string>(); param_value.empty())
+                return tl::make_unexpected(
+                    fmt::format("The parameter '{}' cannot be empty.", parameter.get_name()));
+            break;
+        default:
+            if (auto param_value = parameter.get_value<std::vector<T>>(); param_value.empty())
+                return tl::make_unexpected(
+                    fmt::format("The parameter '{}' cannot be empty.", parameter.get_name()));
+    }
+    return {};
+}
+
+/**
+ * @brief Are all elements of parameter within the bounds (inclusive)?
+ * @pre rclcpp::Parameter must be an array type
+ * @tparam T the interior type of array; Example: for parameter type double_array, T = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto element_bounds(rclcpp::Parameter const& parameter, T const& lower,
+                                  T const& upper) -> tl::expected<void, std::string> {
+    auto const& param_value = parameter.get_value<std::vector<T>>();
+    for (auto val : param_value)
+        if (val < lower || val > upper)
+            return tl::make_unexpected(
+                fmt::format("Invalid value '{}' for parameter '{}'. Required bounds: [{}, {}]", val,
+                            parameter.get_name(), lower, upper));
+    return {};
+}
+
+/**
+ * @brief Are all elements of parameter greater than lower bound?
+ * @pre rclcpp::Parameter must be an array type
+ * @tparam T the interior type of array; Example: for parameter type double_array, T = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto lower_element_bounds(rclcpp::Parameter const& parameter, T const& lower)
```

Comment:
```suggestion
[[nodiscard]] auto all_greater_than(rclcpp::Parameter const& parameter, T const& lower)
```
Does this name better capture the intent of the function?

Reply:
Can we do this after we migrate gpl to using RSL then we can go back and look at these names.